### ISSUE TITLE
Expose SamlResponse parser

### ISSUE THREAD

**Issue by mishawakerman** [bergie/passport-saml#365]
I'm still wrapping my head around some of this so correct me if any base assumptions are wrong :)

I'm looking to implement a `MultiSamlStrategy` that allows our clients to use their respective SSO services. Rather than implement a seperate callback route for each client, I'm wanting to re-use the same `POST /saml/callback` for all all of the strategies.

My current implementation looks to add the `samlOption` to the request object via middleware prior to the passport middleware:

```
router.get('/login-sso',
    emailToSamlOptions(req => req.query.email), // sets req.sso based on email address
    passport.authenticate('multi-saml', { failureRedirect: '/login-error' })
);

router.post('/saml/callback',
    emailToSamlOptions((req) => /* TODO - get the email address from req... */),
    passport.authenticate('multi-saml', { failureRedirect: '/login-error'  }),
    (req, res) => {
        res.redirect('/');
    }
);
```

And my `MultiSamlStrategy` looks like this:

```
export const MultiSamlStrategy = new Strategy(
    {
        name: 'multi-saml',
        // issuer: config.get('client_url'), TODO: for some reason putting the issuer here doesn't work
        getSamlOptions: (req, done) => {
            const options = req.sso;
            if (!options) {
                return done(new Error('no sso options found on request'));
            }
            return done(null, {
                ...options,
                issuer: config.get('client_url'),
                path: '/saml/callback',
            });
        }
    },
    (profile, done) => {
        const { email } = profile;
        findOrCreateUser(email).then(user => done(null, user).catch(err => don(err, null))
    }
);
```

Long story short, in order to get the email address from the SAML response, I have to replicate the `xml2js` parsing that occurs in the library:
```
...
var xml = new Buffer(container.SAMLRequest, 'base64').toString('utf8');
var parserConfig = {
  explicitRoot: true,
  tagNameProcessors: [xml2js.processors.stripPrefix]
};
var parser = new xml2js.Parser(parserConfig);
  parser.parseString(xml, function (err, doc) {
    if (err) {
      return callback(err);
    }
...
```

I've also noticed that the pattern above appears 3-4 times. I'd like to propose that we move the parsing to it's own method and expose this on the SAML export. I'm happy to submit this PR if approved.

**Comment by markstos** [comment]
Yes, splitting out the passport-specific code and the SAML-processing code is welcome, as would be a more promise-friendly interface. 

Closing as duplicate of #295. Also see the conversation here:
 https://github.com/bergie/passport-saml/pull/248#issuecomment-424386991
